### PR TITLE
[ADAM-1257] Add program record support for alignment/fragment files.

### LINKED ISSUE #1257: Removing ProgramRecords info when saving data to sam/bam?
Hi,

Before ADAM,I got PG info in the sam file:
image

After ADAM, the output sam/bam file lost the PG info
image

### PR DESCRIPTION
**Pull request by fnothaft**
Resolves #1257. Depends on https://github.com/bigdatagenomics/adam/pull/1523. WIP:

- [x] ~~Need to connect to `RecordGroup`~~ deferring for now
- [x] Need to look at `org.bdgenomics.adam.models.ProgramGroup`.

### CODE REVIEW

**Comment by fnothaft** [comment]
Also @heuermh the VCF side of this is less well defined, since VCF doesn't specify what the processing description header line should be (e.g., GATK writes out a `GATKProcessingDescription` header line, or some such). Thoughts? I'm thinking that we should handle processing description to VCF in this PR, but punt VCF -> processing description until later.

**Comment by heuermh** [comment]
> the VCF side of this is less well defined, since VCF doesn't specify what the processing description header line should be (e.g., GATK writes out a `GATKProcessingDescription` header line, or some such). Thoughts? I'm thinking that we should handle processing description to VCF in this PR, but punt VCF -> processing description until later.

+1, add our own header line key(s)

**Comment by fnothaft** [comment]
Ping for review.

**Comment by fnothaft** [comment]
Jenkins, test this please.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala [hunk 1]
```diff
@@ -143,6 +144,23 @@ object ADAMContext {
...
   // Add implicits for the rich adam objects
   implicit def recordToRichRecord(record: AlignmentRecord): RichAlignmentRecord = new RichAlignmentRecord(record)
+
+  /**
+   * Builds a program description from a htsjdk program record.
+   *
+   * @param record Program record to convert.
+   * @return Returns an Avro formatted program record.
+   */
+  private[adam] def convertSAMProgramRecord(
```
Comment (devin-petersohn) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: I am curious about the permissions here. Can we be stricter (`private[rdd]`)?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala [hunk 2]
```diff
@@ -186,6 +204,16 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
   }
 
   /**
+   * @param samHeader The header to extract processing lineage from.
+   * @return Returns the dictionary converted to an Avro model.
+   */
+  private[rdd] def loadBamPrograms(
+    samHeader: SAMFileHeader): Seq[ProcessingStep] = {
+    val pgs = samHeader.getProgramRecords().toSeq
+    pgs.map(ADAMContext.convertSAMProgramRecord)
```
Comment (devin-petersohn) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: Is there a namespace collision that requires you to call this as `ADAMContext.convertSAMProgramRecord`?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala (hunk 1 above)
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: Nope; it is needed in `org.bdgenomics.adam.models`.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala (hunk 2 above)
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: I moved the method into the companion object; this is needed to compile.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala (hunk 1 above)
Comment (devin-petersohn) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: I see it, my mistake.

**Comment by fnothaft** [comment]
Rebased. Ping for review @heuermh.

**Comment by fnothaft** [comment]
Rebased and cleared conflicts.

**Comment by fnothaft** [comment]
Jenkins, test this please.

**Comment by fnothaft** [comment]
Jenkins, test this please.

**Comment by fnothaft** [comment]
Jenkins, retest this please.

**Comment by fnothaft** [comment]
Ping for merge @heuermh @devin-petersohn

**Comment by heuermh** [comment]
LGTM, still failing Jenkins though

**Comment by fnothaft** [comment]
Jeez, I missed that somehow... Let me fix.

**Comment by fnothaft** [comment]
Jenkins, test this please.

**Comment by fnothaft** [comment]
OK, this is good for merge now.

**Comment by devin-petersohn** [comment]
Thanks @fnothaft!

**Pull request merged**